lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -458,3 +461,72 @@ void SupplementaryOutputPathsComputer::deriveModulePathParameters(
   mainOutputIfUsable =
       canUseMainOutputForModule && !OutputFiles.empty() ? OutputFiles[0] : "";
 }
+
+static SupplementaryOutputPaths
+createFromTypeToPathMap(const TypeToPathMap *map) {
+  SupplementaryOutputPaths paths;
+  if (!map)
+    return paths;
+  const std::pair<types::ID, std::string &> typesAndStrings[] = {
+      {types::TY_ObjCHeader, paths.ObjCHeaderOutputPath},
+      {types::TY_SwiftModuleFile, paths.ModuleOutputPath},
+      {types::TY_SwiftModuleDocFile, paths.ModuleDocOutputPath},
+      {types::TY_Dependencies, paths.DependenciesFilePath},
+      {types::TY_SwiftDeps, paths.ReferenceDependenciesFilePath},
+      {types::TY_SerializedDiagnostics, paths.SerializedDiagnosticsPath},
+      {types::TY_ModuleTrace, paths.LoadedModuleTracePath},
+      {types::TY_TBD, paths.TBDPath}};
+  for (const std::pair<types::ID, std::string &> &typeAndString :
+       typesAndStrings) {
+    auto const out = map->find(typeAndString.first);
+    typeAndString.second = out == map->end() ? "" : out->second;
+  }
+  return paths;
+}
+
+Optional<std::vector<SupplementaryOutputPaths>>
+SupplementaryOutputPathsComputer::readSupplementaryOutputFileMap() const {
+  if (Arg *A = Args.getLastArg(options::OPT_emit_objc_header_path,
+                               options::OPT_emit_module_path,
+                               options::OPT_emit_module_doc_path,
+                               options::OPT_emit_dependencies_path,
+                               options::OPT_emit_reference_dependencies_path,
+                               options::OPT_serialize_diagnostics_path,
+                               options::OPT_emit_loaded_module_trace_path,
+                               options::OPT_emit_tbd_path)) {
+    Diags.diagnose(
+        SourceLoc(),
+        diag::
+            error_cannot_have_supplementary_outputs_with_supplementary_ouputs_file_map,
```

Comment:
Nitpick: I think you might as well shorten the name at this point. It's not buying us anything to repeat "supplementary outputs", for example. (Particularly with a typo in it ;-) )

Reply:
Shortened. And I continue to be impressed by your ability to spot those details!

Reply:
Well, it took me three times around to see the typo!